setup.py

```diff
@@ -67,39 +70,157 @@
-extras["all"] = extras["serving"] + ["tensorflow", "torch"]
-
-extras["sentencepiece"] = ["sentencepiece!=0.1.92"]
-extras["retrieval"] = ["faiss-cpu", "datasets"]
-extras["testing"] = ["pytest", "pytest-xdist", "timeout-decorator", "parameterized", "psutil"] + extras["retrieval"]
-# sphinx-rtd-theme==0.5.0 introduced big changes in the style.
-extras["docs"] = ["recommonmark", "sphinx", "sphinx-markdown-tables", "sphinx-rtd-theme==0.4.3", "sphinx-copybutton"]
-extras["quality"] = ["black >= 20.8b1", "isort >= 5.5.4", "flake8 >= 3.8.3"]
-extras["dev"] = extras["testing"] + extras["quality"] + extras["ja"] + ["scikit-learn", "tensorflow", "torch", "sentencepiece!=0.1.92"]
+extras["tf-deps"] = to_list("""
+    keras2onnx
+    onnxconverter-common
+    # onnxconverter-common @ git+git://github.com/microsoft/onnxconverter-common.git@f64ca15989b6dc95a1f3507ff6e4c395ba12dff5#egg=onnxconverter-common,
+    # keras2onnx @ git+git://github.com/onnx/keras-onnx.git@cbdc75cb950b16db7f0a67be96a278f8d2953b48#egg=keras2onnx
+""")
+
+# keras2onnx and onnxconverter-common version is specific through a commit until 1.7.0 lands on pypi
+extras["tf-main"] = to_list("""
+    tensorflow>=2.0
+""")
+
+extras["tf-cpu-main"] = to_list("""
+    tensorflow-cpu>=2.0
+""")
+
+extras["tf"] = combine_targets(to_list("""
+    tf-deps
+    tf-main
+"""))
+
+extras["tf-cpu"] = combine_targets(to_list("""
+    tf-cpu-main
+    tf-deps
+"""))
+
+extras["torch"] = to_list("""
+    torch>=1.0
+""")
+
+extras["flax"] = to_list("""
+    flax==0.2.2
+    jax>=0.2.0
+    jaxlib==0.1.55
+""")
+if os.name == "nt":  # windows
+    extras["flax"] = [] # jax is not supported on windows
+
+extras["onnxruntime"] = to_list("""
+    onnxruntime>=1.4.0
+    onnxruntime-tools>=1.4.2
+""")
+
+extras["serving"] = to_list("""
+    fastapi
+    pydantic
+    starlette
+    uvicorn
+""")
+
+extras["sentencepiece"] = to_list("""
+    sentencepiece!=0.1.92
+""")
+
+extras["retrieval"] = to_list("""
+    datasets
+    faiss-cpu
+""")
+
+extras["testing-base"] = to_list("""
+    parameterized
+    psutil
+    pytest
+    pytest-xdist
+    timeout-decorator
+""")
+
+extras["testing"] = combine_targets(to_list("""
+    testing-base
+    retrieval
+"""))
+
+extras["docs"] = to_list("""
+    recommonmark
+    sphinx
+    sphinx-copybutton
+    sphinx-markdown-tables
+    sphinx-rtd-theme==0.4.3 # sphinx-rtd-theme==0.5.0 introduced big changes in the style
+""")
+
+extras["tokenizers"] = to_list("""
+    tokenizers==0.9.2
+""")
+
+extras["quality"] = to_list("""
+    black>=20.8b1
+    flake8>=3.8.3
+    isort>=5.5.4
+""")
+
+extras["dev"] = combine_targets(to_list("""
+    docs
+    flax
+    ja
+    quality
+    sentencepiece
+    sklearn
+    testing
+    tf
+    torch
+"""))
+
+extras["all"] = combine_targets(to_list("""
+    flax
+    sentencepiece
+    tf
+    tokenizers
+    torch
+"""))
+
+# debug 
+if 0:
+    from pprint import pprint
+    pprint(extras)
```

Comment:
Debug necessary?

Reply:
For now - so that if we tweak, it's easy to validate that everything is in place still.